### PR TITLE
[geographiclib] Update to 1.50.1

### PR DESCRIPTION
**Pull request by cffk**
Upgrade to geographiclib 1.50.1

- What does your PR fix? Fixes #11687, a stalled upgrade.

- Which triplets are supported/not supported? Tested on Linux, MacOS, Windwos

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)? Yes.

I am the author of GeographicLib and am willing to take "ownership" of this vcpkg port.

### CODE REVIEW

**Comment by cffk** [comment]
Well, I'm temporarily defeated by the `*_uwp` failures.  The error logs are completely opaque and I have no idea what `uwp` even means.  Any assistance would be appreciated.

**Comment by strega-nil** [comment]
@cffk `uwp` is "Universal Windows Platform"; it's this kinda thing https://docs.microsoft.com/en-us/windows/uwp/get-started/universal-application-platform-guide. They used to be the only thing you could publish to the windows store.

**Comment by cffk** [comment]
I tracked down the problem with `uwp`.  Evidently, there's a problem creating executables even if they are only used for testing and are not installed.  So this PR is ready to go now.

**Comment by cffk** [comment]
@strega-nil The issue with `uwp` and compiling executable programs is the error, e.g.,

    error APPX0703: Manifest references file 'Planimeter.exe'
    which is not part of the payload

The ideal solution would be figure out how to resolve this error.  Failing that, what's the cleanest way for `cmake` to figure out it's compiling for `uwp` so I can skip the building of executables for this platform?

**Comment by cffk** [comment]
OK, this PR is ready for merging!

I fixed the `uwp` problem by not building the tools for `uwp`.  But it would be good to know if it is possible to avoid this restriction.

File: ports/geographiclib/portfile.cmake [hunk 1]
```diff
@@ -1,46 +1,55 @@
...
-    PATCHES
-       remove-tools-and-fix-version.patch
-       fix-usage.patch
+vcpkg_from_sourceforge (
+  OUT_SOURCE_PATH SOURCE_PATH
+  REPO geographiclib
+  REF distrib
+  FILENAME "GeographicLib-1.50.1.tar.gz"
+  SHA512 1db874f30957a0edb8a1df3eee6db73cc993629e3005fe912e317a4ba908e7d7580ce483bb0054c4b46370b8edaec989609fb7e4eb6ba00c80182db43db436f1
+  PATCHES cxx-library-only.patch
```
Comment (NancyLi1013) [inline ports/geographiclib/portfile.cmake:7]:
```suggestion
vcpkg_from_sourceforge (
    OUT_SOURCE_PATH SOURCE_PATH
    REPO geographiclib
    REF distrib
    FILENAME "GeographicLib-1.50.1.tar.gz"
    SHA512 1db874f30957a0edb8a1df3eee6db73cc993629e3005fe912e317a4ba908e7d7580ce483bb0054c4b46370b8edaec989609fb7e4eb6ba00c80182db43db436f1
    PATCHES cxx-library-only.patch
```

File: ports/geographiclib/portfile.cmake [hunk 2]
```diff
@@ -1,46 +1,55 @@
...
+else ()
+  set (LIB_TYPE "STATIC")
+endif ()
+
+vcpkg_configure_cmake (
+  SOURCE_PATH ${SOURCE_PATH}
+  OPTIONS
+  "-DGEOGRAPHICLIB_LIB_TYPE=${LIB_TYPE}"
+  "-DSKIP_TOOLS=${SKIP_TOOLS}"
+  PREFER_NINJA # Disable this option if project cannot be built with Ninja
```
Comment (NancyLi1013) [inline ports/geographiclib/portfile.cmake:27]:
```suggestion
vcpkg_configure_cmake (
    SOURCE_PATH ${SOURCE_PATH}
    OPTIONS
        "-DGEOGRAPHICLIB_LIB_TYPE=${LIB_TYPE}"
        "-DSKIP_TOOLS=${SKIP_TOOLS}"
        PREFER_NINJA # Disable this option if project cannot be built with Ninja
```

File: ports/geographiclib/portfile.cmake [hunk 3]
```diff
@@ -1,46 +1,55 @@
...
+file (INSTALL ${SOURCE_PATH}/LICENSE.txt
+  DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT}
+  RENAME copyright)
+
+# Install usage
+configure_file (${CMAKE_CURRENT_LIST_DIR}/usage
+  ${CURRENT_PACKAGES_DIR}/share/${PORT}/usage @ONLY)
+
+# Post-build test for cmake libraries
+vcpkg_test_cmake(PACKAGE_NAME GeographicLib)
```
Comment (NancyLi1013) [inline ports/geographiclib/portfile.cmake:55]:
```suggestion
```
This function cannot work now. So we should disable or remove it.

File: ports/geographiclib/vcpkg.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+  "name": "geographiclib",
+  "version-string": "1.50.1",
+  "homepage": "https://geographiclib.sourceforge.io",
+  "description": "a small set of C++ classes for performing conversions between geographic, UTM, UPS, MGRS, geocentric, and local cartesian coordinates, for gravity (e.g., EGM2008), geoid height, and geomagnetic field (e.g., WMM2010) calculations, and for solving geodesic problems.",
+  "dependencies": [],
+  "default-features": [],
+  "features": []
```
Comment (NancyLi1013) [inline ports/geographiclib/vcpkg.json:8]:
```suggestion
```
These fields are not used in this port. So we can remove them here.

**Review by NancyLi1013 — DISMISSED** [review]

File: ports/geographiclib/portfile.cmake (hunk 1 above)
Comment (cffk) [reply ports/geographiclib/portfile.cmake]: Done

File: ports/geographiclib/portfile.cmake (hunk 2 above)
Comment (cffk) [reply ports/geographiclib/portfile.cmake]: Done.  But I did *not* align `PREFER_NINJA` as though it was an option.

File: ports/geographiclib/portfile.cmake (hunk 3 above)
Comment (cffk) [reply ports/geographiclib/portfile.cmake]: Done

File: ports/geographiclib/vcpkg.json (hunk 4 above)
Comment (cffk) [reply ports/geographiclib/vcpkg.json]: Done

File: ports/geographiclib/portfile.cmake [hunk 5]
```diff
@@ -1,46 +1,52 @@
...
-        # OPTIONS -DUSE_THIS_IN_ALL_BUILDS=1 -DUSE_THIS_TOO=2
-        # OPTIONS_RELEASE -DOPTIMIZE=1
-        # OPTIONS_DEBUG -DDEBUGGABLE=1
-    )
-endif()
+if (TARGET_TRIPLET MATCHES ".*-uwp")
+  set (SKIP_TOOLS ON)
+else ()
+  set (SKIP_TOOLS OFF)
+endif ()
```
Comment (strega-nil) [inline ports/geographiclib/portfile.cmake:14]: I think the thing to do here is use features instead; something like https://github.com/microsoft/vcpkg/blob/master/ports/microsoft-signalr/CONTROL#L8.

Then, you'd just do

```cmake
vcpkg_check_features(
  OUT_FEATURE_OPTIONS FEATURE_OPTIONS
  INVERTED_FEATURES
    "tools" SKIP_TOOLS
)
```

**Comment by johnco3** [comment]
Tried to install the pull request after building locally but it failed.

```
C:\Users\johnc\main\extlibs\vcpkg1>.\vcpkg.exe install geographiclib
Computing installation plan...
Error: while loading geographiclib:
Error: C:/Users/johnc/main/extlibs/vcpkg1/ports/geographiclib/vcpkg.json:6:1: Trailing comma in an object
   on expression: {
  "name": "geographiclib",
  "version-string": "1.50.1",
  "homepage": "https://geographiclib.sourceforge.io",
  "description": "a small set of C++ classes for performing conversions between geographic, UTM, UPS, MGRS, geocentric, and local cartesian coordinates, for gravity (e.g., EGM2008), geoid height, and geomagnetic field (e.g., WMM2010) calculations, and for solving geodesic problems.",
}
                                                                                                                                                                                                                                                                                                                                                                                                                                ^

Error: Failed to load port from geographiclib

C:\Users\johnc\main\extlibs\vcpkg1>.\vcpkg.exe search geographiclib --debug
[DEBUG] Feature flag 'binarycaching' unset
[DEBUG] Feature flag 'manifests' unset
[DEBUG] Feature flag 'compilertracking' unset
[DEBUG] Using vcpkg-root: C:\Users\johnc\main\extlibs\vcpkg1
[DEBUG] Using installed-root: C:\Users\johnc\main\extlibs\vcpkg1\installed
[DEBUG] Using buildtrees-root: C:\Users\johnc\main\extlibs\vcpkg1\buildtrees
[DEBUG] Using downloads-root: C:\Users\johnc\main\extlibs\vcpkg1\downloads
[DEBUG] Using packages-root: C:\Users\johnc\main\extlibs\vcpkg1\packages
[DEBUG] Using scripts-root: C:\Users\johnc\main\extlibs\vcpkg1\scripts
Error: while loading geographiclib:
Error: C:/Users/johnc/main/extlibs/vcpkg1/ports/geographiclib/vcpkg.json:6:1: Trailing comma in an object
   on expression: {
  "name": "geographiclib",
  "version-string": "1.50.1",
  "homepage": "https://geographiclib.sourceforge.io",
  "description": "a small set of C++ classes for performing conversions between geographic, UTM, UPS, MGRS, geocentric, and local cartesian coordinates, for gravity (e.g., EGM2008), geoid height, and geomagnetic field (e.g., WMM2010) calculations, and for solving geodesic problems.",
}
                                                                                                                                                                                                                                                                                                                                                                                                                                ^

If your library is not listed, please open an issue at and/or consider making a pull request:
    https://github.com/Microsoft/vcpkg/issues
[DEBUG] C:\Users\johnc\main\extlibs\vcpkg1\toolsrc\src\vcpkg\commands.search.cpp(146)
[DEBUG] Exiting after 341452 us (332706 us)

C:\Users\johnc\main\extlibs\vcpkg1>

```

File: ports/geographiclib/vcpkg.json [hunk 6]
```diff
@@ -0,0 +1,6 @@
+{
+  "name": "geographiclib",
+  "version-string": "1.50.1",
+  "homepage": "https://geographiclib.sourceforge.io",
+  "description": "a small set of C++ classes for performing conversions between geographic, UTM, UPS, MGRS, geocentric, and local cartesian coordinates, for gravity (e.g., EGM2008), geoid height, and geomagnetic field (e.g., WMM2010) calculations, and for solving geodesic problems.",
```
Comment (strega-nil) [inline ports/geographiclib/vcpkg.json:5]: And here, you'd do:

```json
{
    "default-features": [ "default-features" ],
    "features": [
        {
            "name": "default-features",
            "description": "Features installed by default",
            "dependencies": [
                {
                    "name": "geographiclib",
                    "features": [ "tools" ],
                    "platform": "!uwp"
                }
            ]
        },
        {
            "name": "tools",
            "description": "Build the geographiclib tools"
        }
    ]
}
```

**Review by strega-nil — CHANGES_REQUESTED** [review]

File: ports/geographiclib/vcpkg.json (hunk 6 above)
Comment (strega-nil) [reply ports/geographiclib/vcpkg.json:5]: (note to self, I really, really wish there was a nicer way to do this "default features only on a certain platform" thing.)

**Comment by cffk** [comment]
@johnco3 Just edit vcpkg.json to remove the last comma and you should be OK.  I'm not making this change myself, because I've got other changes to do to this file.

File: ports/geographiclib/portfile.cmake (hunk 5 above)
Comment (cffk) [reply ports/geographiclib/portfile.cmake]: OK, thanks.  I've pushed this change.

File: ports/geographiclib/vcpkg.json (hunk 6 above)
Comment (cffk) [reply ports/geographiclib/vcpkg.json:5]: Done...

**Review by strega-nil — APPROVED** [review]

**Comment by cffk** [comment]
@johnco3 I've pushed changes which should fix your problem...

File: ports/geographiclib/usage [hunk 7]
```diff
@@ -1,8 +1,4 @@
...
-    #dynamic
-    target_link_libraries(main PRIVATE GeographicLib GeographicLib::GeographicLib)
-    #static
-    target_link_libraries(main PRIVATE GeographicLib_STATIC GeographicLib::GeographicLib_STATIC)
-    
\ No newline at end of file
+The package @PORT@:@TARGET_TRIPLET@ provides CMake targets:
+
+    find_package (GeographicLib CONFIG REQUIRED)
+    target_link_libraries (main PRIVATE ${GeographicLib_LIBRARIES})
```
Comment (ras0219-msft) [inline ports/geographiclib/usage:4]: We want to encourage using CMake targets as much as possible, so perhaps this should be
```suggestion
    target_link_libraries (main PRIVATE GeographicLib::GeographicLib)
```

**Review by ras0219-msft — DISMISSED** [review]
LGTM for merge once the two comments are addressed!

File: ports/geographiclib/vcpkg.json [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+      "dependencies": [
+        {
+          "name": "geographiclib",
+          "features": [
+            "tools"
+          ],
+          "platform": "!uwp"
+        }
+      ]
+    },
```
[the remark this answers is not in the document]
Comment (cffk) [reply ports/geographiclib/vcpkg.json:22]: I'm going to push back here.  Many (most?) of the people you use GeographicLib use it *only* via the tools provided with it (or with a combination of the library and the tools).  If `vcpkg` doesn't provide these tools, then these people are forced to find some other way of installing GeograpicLib.  This doesn't make any sense to me.

File: ports/geographiclib/usage (hunk 7 above)
Comment (cffk) [reply ports/geographiclib/usage:4]: The cmake expression `${GeographicLib_LIBRARIES}` *is* a target.  But the use via a cmake variable ensure a graceful transition from non-namespace and namespace conventions.  This transition will be complete in a year or two.  This is also consistent with the current [documentation](https://geographiclib.sourceforge.io/html/start.html) provided for the end user.
In any case, the use of a variable following a pretty standard convention ${Package_LIBRARIES} is much easier for the end user than dealing with the various (and changeable) conventions on the target names.  For instance, contrast ZLIB::ZLIB with GSL::gsl.

File: ports/geographiclib/vcpkg.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply ports/geographiclib/vcpkg.json:22]: That's why there is still a tools feature, which is as easy to acquire as `vcpkg install geographiclib[tools]`. However, the common _transitive_ use doesn't need the tool (I use a lib which uses geographiclib), so it doesn't make sense to opt in to building them for users who haven't explicitly named the library.

In the case of `shapelib`, similar logic applies and I'm sorry that we don't have 100% adherence to this policy in master yet.

We'd be happy to review a PR adding an off-by-default tools feature to `gdal`!

[the remark this answers is not in the document]
Comment (cffk) [reply ports/geographiclib/vcpkg.json:22]: OK, sorry, I misunderstood.  Let me take a look.

[the remark this answers is not in the document]
Comment (Mistrical) [reply ports/geographiclib/vcpkg.json:22]: I might be in the minority here, but I do use GeographicLib _purely as a library_ for geographic coordinate conversions in my C++ application, and have never needed to build/use the utility programs.

Given that vcpkg's main prupose is to build and install C/C++ libraries for developers to use, it makes sense to exclude the tools from the default build, and make the tools optional. Devs usually don't want to spend time building tools that they don't need.

[the remark this answers is not in the document]
Comment (cffk) [reply ports/geographiclib/vcpkg.json]: OK, I made `tools` a non-default feature and I put a test in portfile.cmake to disallow using `tools` with `uwp`.

[the remark this answers is not in the document]
Comment (cffk) [reply ports/geographiclib/vcpkg.json]: > I might be in the minority here, but I do use GeographicLib _purely as a library_ for geographic coordinate conversions in my C++ application, and have never needed to build/use the utility programs.
> 
> Given that vcpkg's main prupose is to build and install C/C++ libraries for developers to use, it makes sense to exclude the tools from the default build, and make the tools optional. Devs usually don't want to spend time building tools that they don't need.

Fair enough.  However, there's plenty of need for the tools, so I wish the ports in `vcpkg` provided these more routinely.

**Review by ras0219 — APPROVED** [review]

**Comment by cffk** [comment]
There's still a problem with the `tools` because `find_package` doesn't find them where they were installed (because they got copied to `tools/geographiclib`).  I'll duct-tape around this problem.  However, it points out that the `vcpkg` after-the-face fixups run somewhat counter to the `cmake` way of doing things.

**Review by NancyLi1013 — APPROVED** [review]

**Comment by ras0219-msft** [comment]
Thanks for the new library and responding to our reviews so quickly! :)

**Pull request merged**